google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITGrpcIdempotencyTokenTest.java

```diff
@@ -0,0 +1,168 @@
+
+    List<String> collect = page.streamAll().map(BlobInfo::getName).collect(Collectors.toList());
+    IterableSubject subject = requestAuditing.assertRequestHeader(X_GOOG_GCS_IDEMPOTENCY_TOKEN);
+    assertAll(
+        () -> assertThat(collect).hasSize(2),
+        () -> assertThat(collect).containsExactlyElementsIn(expectedNamess),
+        () -> subject.hasSize(2));
+  }
+
+  @Test
+  public void readObject() throws Exception {
+    byte[] expected = DataGenerator.base64Characters().genBytes(512 * 1024 + 45);
+    String expectedXxd = xxd(expected);
+
+    Blob gen1 =
+        storage.create(
+            BlobInfo.newBuilder(bucket, generator.randomObjectName()).build(),
+            expected,
+            BlobTargetOption.doesNotExist());
+
+    requestAuditing.clear();
+    byte[] actual = storage.readAllBytes(gen1.getBlobId(), BlobSourceOption.generationMatch());
+    IterableSubject subject = requestAuditing.assertRequestHeader(X_GOOG_GCS_IDEMPOTENCY_TOKEN);
+    String actualXxd = xxd(actual);
+
+    assertAll(() -> subject.hasSize(1), () -> assertThat(actualXxd).isEqualTo(expectedXxd));
+  }
+
+  @Test
+  public void directUpload() throws Exception {
+    byte[] expected = DataGenerator.base64Characters().genBytes(512 * 1024 + 45);
+    String expectedXxd = xxd(expected);
+
+    BlobInfo info = BlobInfo.newBuilder(bucket, generator.randomObjectName()).build();
+    requestAuditing.clear();
+    Blob gen1 = storage.create(info, expected, BlobTargetOption.doesNotExist());
+    IterableSubject subject = requestAuditing.assertRequestHeader(X_GOOG_GCS_IDEMPOTENCY_TOKEN);
+
+    byte[] actual = storage.readAllBytes(gen1.getBlobId(), BlobSourceOption.generationMatch());
+    String actualXxd = xxd(actual);
+
+    assertAll(() -> subject.hasSize(1), () -> assertThat(actualXxd).isEqualTo(expectedXxd));
+  }
+
+  @Test
+  public void resumableUpload() throws Exception {
+    byte[] expected = DataGenerator.base64Characters().genBytes(512 * 1024 + 45);
+    String expectedXxd = xxd(expected);
+
+    BlobInfo info = BlobInfo.newBuilder(bucket, generator.randomObjectName()).build();
+    try (WriteChannel writer = storage.writer(info, BlobWriteOption.doesNotExist())) {
+      writer.setChunkSize(256 * 1024);
+      writer.write(ByteBuffer.wrap(Arrays.copyOfRange(expected, 0, 256 * 1024)));
+      writer.write(ByteBuffer.wrap(Arrays.copyOfRange(expected, 256 * 1024, 512 * 1024)));
+      writer.write(ByteBuffer.wrap(Arrays.copyOfRange(expected, 512 * 1024, expected.length)));
+    }
+    IterableSubject subject = requestAuditing.assertRequestHeader(X_GOOG_GCS_IDEMPOTENCY_TOKEN);
+
+    byte[] actual = storage.readAllBytes(info.getBlobId());
+    String actualXxd = xxd(actual);
+
+    // We expect 4 distinct requests:
+    //  1. start resumable session
+    //  2. PUT first 256KiB
+    //  3. PUT second 256KiB
+    //  4. Finalize session and put final 45B
```

Comment:
Just figured I'd ask because I am not great at following some of this testing code -- does the test assert that each of these 4 gets its own request ID, and that all retries for each request have the same ID?

Reply:
Mostly correct, we are asserting there is a distinct token for each of the 4 requests that are sent, however we are not inducing any failures in these tests to verify tokens are maintained for the same request across multiple attempts. Instead we are waiting on https://github.com/googleapis/storage-testbench/issues/486 for the retry enforcement piece.